gamemodes/terrortown/gamemode/shared/sh_door.lua

```diff
@@ -0,0 +1,184 @@
+-- @internal
+-- @realm shared
+function door.SetUp()
+	local all_ents = ents.GetAll()
+
+	for i = 1, #all_ents do
+		local ent = all_ents[i]
+
+		if not ent:IsDoor() then continue end
+
+		door.__doors[#door.__doors + 1] = ent
+
+		-- set up synced states if on server
+		if CLIENT then continue end
+
+		ent:SetNWBool("ttt2_door_locked", ent:GetInternalVariable("m_bLocked") or false)
+		ent:SetNWBool("ttt2_door_forceclosed", ent:GetInternalVariable("forceclosed") or false)
+		ent:SetNWBool("ttt2_door_open", ent:InternalIsDoorOpen() or false)
+	end
+end
+
+---
+-- Returns all valid door entities found on a map
+-- @return table A table of door entities
+-- @realm shared
+function door.GetAll()
+	return door.__doors
+end
+
+---
+-- Called when a map I/O event occurs.
+-- @param Entity ent Entity that receives the input
+-- @param string input The input name. Is not guaranteed to be a valid input on the entity.
+-- @param Entity activator Activator of the input
+-- @param Entity caller Caller of the input
+-- @param any data Data provided with the input
+-- @return boolean Return true to prevent this input from being processed.
+-- @ref https://wiki.facepunch.com/gmod/GM:AcceptInput
+-- @hook
+-- @realm shared
+function GM:AcceptInput(ent, name, activator, caller, data)
+	if not IsValid(ent) or not ent:IsDoor() then return end
+
+	if name == "lock" then
+		-- we expect the door to be locked now, but we check the real state after a short
+		-- amount of time to be sure
+		ent:SetNWBool("ttt2_door_locked", true)
+
+		-- check if the assumed state was correct
+		timer.Create("ttt2_recheck_door_lock_" .. ent:EntIndex(), 1, 1, function()
+			if not IsValid(ent) then return end
+
+			ent:SetNWBool("ttt2_door_locked", ent:GetInternalVariable("m_bLocked") or false)
+		end)
+	elseif name == "unlock" then
+		-- we expect the door to be unlocked now, but we check the real state after a short
+		-- amount of time to be sure
+		ent:SetNWBool("ttt2_door_locked", false)
+
+		-- check if the assumed state was correct
+		timer.Create("ttt2_recheck_door_unlock_" .. ent:EntIndex(), 1, 1, function()
+			if not IsValid(ent) then return end
+
+			ent:SetNWBool("ttt2_door_locked", ent:GetInternalVariable("m_bLocked") or false)
+		end)
+	elseif name == "Use" then
+		-- upon triggering the state change of the door, the state does not change
+		-- instanly but after the animation finished. Therefore we calculate an assumned
+		-- value on the fly and check the real state a few seconds later
+		if not ent:IsDoorLocked() then
+			ent:SetNWBool("ttt2_door_open", not ent:GetNWBool("ttt2_door_open", false))
+		end
+
+		-- check if the assumed state was correct
+		timer.Create("ttt2_recheck_door_use_" .. ent:EntIndex(), 2.5, 1, function()
+			if not IsValid(ent) or ent:IsDoorLocked() then return end
+
+			ent:SetNWBool("ttt2_door_open", ent:InternalIsDoorOpen() or false)
+		end)
+	end
+end
+
+
+--- ENTITY EXTENSION STUFF ---
```

Comment:
Both in one file?

Reply:
Yes since they both need access to the same local variables. I could make them public or define them twice. I don't like both approaches

Reply:
Ref. the previous comment, the current solution should be the best solution, you are right, but it‘s unconventional ^^